### PR TITLE
add config deploying Kafka dispatcher in user-ns

### PR DESCRIPTION
**Pull request by lionelvillard**
Fixes https://github.com/knative/eventing-contrib/issues/751

## Proposed Changes

  * Add a KafkaChannel
  *
  *

**Release Note**

```release-note
🎁 A new Kafka channel configuration is provided for deploying the Kafka channel dispatcher
in the same namespace as the resource being reconciled.
```

### CODE REVIEW

File: kafka/channel/config/200-controller-clusterrole.yaml [hunk 1]
```diff
@@ -95,3 +95,15 @@ rules:
...
+    resources:
+      - serviceaccounts
+    verbs:
+      - list
+  - apiGroups:
+      - rbac.authorization.k8s.io
+    resources:
+      - rolebindings
+    verbs:
+      - list
```
Comment (mattmoor) [inline kafka/channel/config/200-controller-clusterrole.yaml]: Add trailing newline:
```suggestion
      - list

```

File: kafka/channel/README.md [hunk 2]
```diff
@@ -17,9 +17,15 @@ topics.
    > Kubernetes. It is also possible to use an off-cluster Apache Kafka
    > installation.
 
+1. Decide on where you want the Kafka Channel Dispatcher to be installed,
+   either in `knative-eventing` (Cluster) or in the same namespace as KafkaChannel
```
Comment (matzew) [inline kafka/channel/README.md]: Not sure, but perhaps `Cluster` -> `cluster scoped` ?

File: kafka/channel/README.md [hunk 3]
```diff
@@ -17,9 +17,15 @@ topics.
...
+   objects (Namespace).
+
 1. Now that Apache Kafka is installed, you need to configure the
    `bootstrapServers` value in the `config-kafka` ConfigMap, located inside the
-   `config/400-kafka-config.yaml` file:
+   `config/400-kafka-config.yaml`.
```
Comment (matzew) [inline kafka/channel/README.md]: > located inside the config/400-kafka-config.yaml.   ?

Not located inste the ... FILE ?

File: kafka/channel/pkg/reconciler/controller/kafkachannel.go [hunk 4]
```diff
@@ -182,21 +203,50 @@ func NewController(
 	// Set up watches for dispatcher resources we care about, since any changes to these
 	// resources will affect our Channels. So, set up a watch here, that will cause
 	// a global Resync for all the channels to take stock of their health when these change.
+
```
Comment (matzew) [inline kafka/channel/pkg/reconciler/controller/kafkachannel.go]: add a 

```
filterFn := ScopedFilter(r.systemNamespace, dispatcherServiceName)
```

and than use 

```
FilterFunc: filterFn,
```

File: kafka/channel/pkg/reconciler/controller/kafkachannel.go [hunk 5]
```diff
@@ -182,21 +203,50 @@ func NewController(
...
+		FilterFunc: r.ScopedFilter(r.systemNamespace, dispatcherServiceName),
 		Handler:    r,
 	})
+
+	if r.dispatcherScope == "namespace" {
+		r.serviceAccountLister = serviceAccountInformer.Lister()
+		r.roleBindingLister = roleBindingInformer.Lister()
+
+		serviceAccountInformer.Informer().AddEventHandler(cache.FilteringResourceEventHandler{
+			FilterFunc: r.ScopedFilter(r.systemNamespace, dispatcherDeploymentName),
```
Comment (matzew) [inline kafka/channel/pkg/reconciler/controller/kafkachannel.go]: same

File: kafka/channel/pkg/reconciler/controller/kafkachannel.go [hunk 6]
```diff
@@ -113,15 +122,21 @@ type Reconciler struct {
...
+	roleBindingLister    rbacv1listers.RoleBindingLister
 	impl                 *controller.Impl
 }
 
 var (
 	deploymentGVK = appsv1.SchemeGroupVersion.WithKind("Deployment")
 	serviceGVK    = corev1.SchemeGroupVersion.WithKind("Service")
+
+	scopeNamespace = "namespace"
+	scopeCluster   = "cluster"
```
Comment (matzew) [inline kafka/channel/pkg/reconciler/controller/kafkachannel.go]: is it more 

`clusterScoped` and `namespaceScoped` ?

File: kafka/channel/pkg/reconciler/controller/kafkachannel.go (hunk 5 above)
Comment (lionelvillard) [reply kafka/channel/pkg/reconciler/controller/kafkachannel.go]: done

File: kafka/channel/config/200-controller-clusterrole.yaml (hunk 1 above)
Comment (mattmoor) [inline kafka/channel/config/200-controller-clusterrole.yaml]: Add trailing newline:
```suggestion
      - list

```

Comment (mattmoor) [inline kafka/channel/config/200-controller-clusterrole.yaml]: Add trailing newline:
```suggestion
      - list

```

Comment (mattmoor) [inline kafka/channel/config/200-controller-clusterrole.yaml]: Add trailing newline:
```suggestion
      - list

```

File: kafka/channel/README.md [hunk 7]
```diff
@@ -29,16 +35,40 @@ topics.
      namespace: knative-eventing
    data:
      # Broker URL. Replace this with the URLs for your kafka cluster,
-     # which is in the format of my-cluster-kafka-bootstrap.my-kafka-namespace:9092.
+     # which is in the format of my-cluster-kafka-bootstrap .my-kafka-namespace:9092.
      bootstrapServers: REPLACE_WITH_CLUSTER_URL
    ```
 
-1. Apply the Kafka config:
+   **namespace-scoped**: set `namespace` to where you are planning to   create `KafkaChannel`
```
Comment (matzew) [inline kafka/channel/README.md]: so, there will be _serveral_ CMs - in different namespaces ?

Comment (lionelvillard) [reply kafka/channel/README.md]: yes

File: kafka/channel/config/200-serviceaccount.yaml [hunk 8]
```diff
@@ -20,15 +20,6 @@ metadata:
   labels:
```
Comment (matzew) [inline kafka/channel/config/200-serviceaccount.yaml]: perhaps we make a split of the remaining two, and move them to 

* 200-controller-serviceaccount.yaml
* 200-webhook-serviceaccount.yaml

File: kafka/channel/pkg/reconciler/controller/resources/dispatcher.go [hunk 9]
```diff
@@ -108,4 +109,17 @@ func makeEnv() []corev1.EnvVar {
 		Name:  "CONFIG_LOGGING_NAME",
 		Value: "config-logging",
 	}}
+
+	if args.DispatcherScope == "namespace" {
```
Comment (matzew) [inline kafka/channel/pkg/reconciler/controller/resources/dispatcher.go]: no update in `dispatcher_test.go`  needed ?

**Pull request merged**